bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/audio/ConvertedInputStream.java

```diff
@@ -0,0 +1,121 @@
+            javax.sound.sampled.AudioFormat.Encoding.ULAW, 8000, 8, 1, 1, 8000, false);
+
+    private AudioInputStream pcmUlawInputStream;
+
+    public ConvertedInputStream(InputStream innerInputStream)
+            throws UnsupportedAudioFormatException, UnsupportedAudioFileException, IOException {
+
+        pcmUlawInputStream = getULAWStream(new BufferedInputStream(innerInputStream));
+    }
+
+    public AudioInputStream getAudioInputStream() {
+        return pcmUlawInputStream;
+    }
+
+    @Override
+    public int read(byte @Nullable [] b) throws IOException {
+        return pcmUlawInputStream.read(b);
+    }
+
+    @Override
+    public int read(byte @Nullable [] b, int off, int len) throws IOException {
+        return pcmUlawInputStream.read(b, off, len);
+    }
+
+    @Override
+    public byte[] readAllBytes() throws IOException {
+        return pcmUlawInputStream.readAllBytes();
+    }
+
+    @Override
+    public byte[] readNBytes(int len) throws IOException {
+        return pcmUlawInputStream.readNBytes(len);
+    }
+
+    @Override
+    public int readNBytes(byte @Nullable [] b, int off, int len) throws IOException {
+        return pcmUlawInputStream.readNBytes(b, off, len);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return pcmUlawInputStream.read();
+    }
+
+    @Override
+    public void close() throws IOException {
+        pcmUlawInputStream.close();
+    }
+
+    /**
+     * Ensure the right ULAW format by converting if necessary (two pass)
+     *
+     * @param originalInputStream a mark/reset compatible stream
+     *
+     * @return A ULAW stream (1 channel, 8000hz, 16 bit signed)
+     * @throws IOException
+     * @throws UnsupportedAudioFileException
+     */
+    private AudioInputStream getULAWStream(InputStream originalInputStream)
+            throws UnsupportedAudioFileException, IOException {
+
+        AudioInputStream audioInputStream = AudioSystem.getAudioInputStream(originalInputStream);
+        AudioFormat format = audioInputStream.getFormat();
+
+        boolean frameRateOk = Math.abs(format.getFrameRate() - 8000) < 1;
+        boolean sampleRateOk = Math.abs(format.getSampleRate() - 8000) < 1;
+
+        if (format.getEncoding().equals(Encoding.ULAW) && format.getChannels() == 1 && frameRateOk && sampleRateOk
+                && format.getFrameSize() == 1 && format.getSampleSizeInBits() == 8) {
+            return audioInputStream;
+        }
+
+        AudioInputStream targetPCMFormat = audioInputStream;
+        if (format.getChannels() != 1 || !frameRateOk || !sampleRateOk || format.getFrameSize() != 2
+                || format.getChannels() != 1 || format.getSampleSizeInBits() != 16) {
+            targetPCMFormat = AudioSystem.getAudioInputStream(INTERMEDIARY_PCM_FORMAT, audioInputStream);
+        }
+
+        return AudioSystem.getAudioInputStream(TARGET_ULAW_FORMAT, targetPCMFormat);
```

Comment:
Catch and rethrow done.

For the conversion to PCM 16 bits : yes, I know it sounds silly (convert to 16 bits then reconvert to 8 bits), but it seems unfortunately mandatory.
I don't remember everything, only that I spend too much time debugging this issue ! There seems to be some code preventing conversion to ULAW if source is not in 16 bits. Don't know why.
It anoys me because it adds latency.
Anyway, I'll had a comment.

I have other sinks that converts sounds. I'm thinking there is maybe a need for some cache, to keep already converted data.
But the API is not designed for this, as it transmits only an InputStream with no additional data that can identify an already played sound. Maybe I can add an "UniquelyIdentified" interface to some AudioStream (those based on File or at least TTS), that will provide optional unique key to identify the sound being played ?

Reply:
> I have other sinks that converts sounds. I'm thinking there is maybe a need for some cache, to keep already converted data.
> But the API is not designed for this, as it transmits only an InputStream with no additional data that can identify an already played sound. Maybe I can add an "UniquelyIdentified" interface to some AudioStream (those based on File or at least TTS), that will provide optional unique key to identify the sound being played ?

Sorry, no idea for that. I can just tell you there is a proposal for a TTS cache being reviewed.